hub/util/class_label.py

```diff
@@ -28,10 +34,106 @@ def convert(samples):
     return convert(samples)
 
 
+def convert_to_hash(samples, hash_label_map):
+    def convert(samples):
+        hashes = []
+        for sample in samples:
+            if isinstance(sample, np.ndarray):
+                sample = sample.tolist()
+            if isinstance(sample, list):
+                hashes_ = convert(sample)
+                hashes.append(hashes_)
+            else:
+                if isinstance(sample, str):
+                    hash_ = hash_str_to_int32(sample)
+                    hash_label_map[hash_] = sample
+                else:
+                    hash_ = sample
+                hashes.append(hash_)
+        return hashes
+
+    return convert(samples)
+
+
+def convert_hash_to_idx(hashes, hash_idx_map):
+    def convert(hashes):
+        idxs = []
+        for hash in hashes:
+            if isinstance(hash, list):
+                idxs_ = convert(hash)
+                idxs.extend(idxs_)
+            else:
+                try:
+                    idx = hash_idx_map[hash]
+                    idxs.append(idx)
+                except KeyError:
+                    idxs.append(hash)
+        return idxs
+
+    return convert(hashes)
+
+
 def convert_to_text(inp, class_names: List[str]):
     if isinstance(inp, np.integer):
         idx = int(inp)
         if idx < len(class_names):
             return class_names[idx]
         return None
     return [convert_to_text(item, class_names) for item in inp]
+
+
+def sync_labels(
+    ds, label_temp_tensors, hash_label_maps, num_workers, scheduler, verbose=True
+):
+    hl_maps = defaultdict(OrderedDict)
+    for map in hash_label_maps:
+        for tensor in map:
+            hl_maps[tensor].update(map[tensor])
+    hash_label_maps = hl_maps
+
+    @hub.compute
+    def upload(
+        hash_tensor,
```

Comment:
rename this to hash_tensor_sample (I first got confused about how we're doing .numpy() on entire tensor) and rename ds_out to samples out